lib/sms/parser.c

```diff
@@ -0,0 +1,169 @@
+ */
+static void convert_to_bytes(char *str, uint32_t str_length, uint8_t* buf, uint16_t buf_length)
+{
+	for (int i = 0; i < str_length; i++) {
+		__ASSERT((i >> 1) <= buf_length, "Too small internal buffer");
+
+		/* Character in an even index will be filled into byte that should be
+		 * initialized to zero to make sure the memory area is clean. */
+		if (!(i % 2)) {
+			buf[i >> 1] = 0;
+		}
+
+		/* The integer value of each character is added into the output buffer into
+		 * index that's half of the characters index in the input buffer as each byte is
+		 * represented by two characters. Division by 2 is done with right shift.
+		 * The first character of the two character set forms the most significant
+		 * 4 bits of the integer and the second character forms the least significant
+		 * 4 bits. This is where the shifting by four comes from based on whether the
+		 * index is even or odd.
+		 */
+		buf[i >> 1] |= (char2int(str[i]) << (4 * !(i % 2)));
+	}
+}
+
+int parser_create(struct parser *parser, struct parser_api *api)
+{
+	memset(parser, 0, sizeof(struct parser));
+
+	parser->api = api;
+
+	parser->data = k_calloc(1, api->data_size());
+	if (parser->data == NULL) {
+		LOG_ERR("Out of memory when creating parser");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+int parser_delete(struct parser *parser)
+{
+	k_free(parser->data);
+	parser->data = NULL;
+
+	return 0;
+}
+
+/**
+ * @brief Process given data with all sub parser of the given parser.
+ * 
+ * @param[in] parser Parser instance which has information on sub parsers that are executed.
+ *
+ * @return Zero if successful, negative value in error cases.
+ */
+static int parser_process(struct parser *parser)
+{
+	int ofs_inc;
+
+	parser_module *parsers = parser->api->get_parsers();
+
+	parser->buf_pos = 0;
+
+	for (int i = 0; i < parser->api->get_parser_count(); i++) {
+		ofs_inc = parsers[i](parser, &parser->buf[parser->buf_pos]);
+
+		if (ofs_inc < 0) {
+			return ofs_inc;
+		}
+
+		parser->buf_pos += ofs_inc;
+
+		/* If we have gone beyond the length of the given data,
+		   we need to return a failure. We don't have issues in
+		   accessing memory beyond parser->buf_size bytes of
+		   parser->buf as the buffer is overly long.
+		   */
+		if (parser->buf_pos > parser->buf_size) {
+			return -EMSGSIZE;
+		}
+	}
+
+	parser->payload_pos = parser->buf_pos;
+
+	return 0;
+}
+
+int parser_process_str(struct parser *parser, char *data)
+{
+	uint16_t length = strlen(data);
+	uint16_t buf_size = length / 2;
+
+	if (buf_size > PARSER_BUF_SIZE) {
+		LOG_ERR("Data length (%d) is bigger than the internal buffer size (%d)",
+			buf_size,
+			PARSER_BUF_SIZE);
+		return -EMSGSIZE;
+	}
+
+	parser->buf_size = buf_size;
+	convert_to_bytes(data, length, parser->buf, PARSER_BUF_SIZE);
+
+	return parser_process(parser);
+}
+
+int parser_get_payload(struct parser *parser, char *buf, uint8_t buf_size)
+{
+	parser_module payload_decoder = parser->api->get_decoder();
+
+	parser->payload = buf;
```

Comment:
Output buffer to the payload decoder is passed using parser->payload. Decided not to refactor this at least at this point.
This would require that we'll have another typedef for payload parsing in addition to parser_module.